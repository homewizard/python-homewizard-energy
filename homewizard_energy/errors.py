"""python-homewizard-energy errors."""


class HomeWizardEnergyException(Exception):
    """Base error for python-homewizard-energy."""


class RequestError(HomeWizardEnergyException):
    """Unable to fulfill request.

    Raised when host or API cannot be reached.
    """


class InvalidUserNameError(HomeWizardEnergyException):
    """Invalid username.

    Raised when username is not valid, too short or too long.
    """


class ResponseError(HomeWizardEnergyException):
    """API responded unexpected."""


class NotFoundError(HomeWizardEnergyException):
    """Request not found.

    Raised when API responds with '404'
    """


class InvalidStateError(HomeWizardEnergyException):
    """Raised when the device is not in the correct state."""


class UnsupportedError(HomeWizardEnergyException):
    """Raised when the device is not supported from this library."""


class DisabledError(HomeWizardEnergyException):
    """Raised when device API is disabled. User has to enable API in app."""


class UnauthorizedError(HomeWizardEnergyException):
    """Raised when request is not authorized."""
