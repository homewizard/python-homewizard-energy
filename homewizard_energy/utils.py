"""Utilities for Python HomeWizard Energy."""

from functools import lru_cache

from awesomeversion import AwesomeVersion


@lru_cache
def get_awesome_version(version: str) -> AwesomeVersion:
    """Return a cached AwesomeVersion object."""
    if version.lower() == "v1":
        return AwesomeVersion("1.0.0")
    return AwesomeVersion(version)
