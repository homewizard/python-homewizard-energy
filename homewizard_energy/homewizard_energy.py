"""HomeWizard Energy base class"""

from __future__ import annotations

import asyncio
from typing import Any

from aiohttp.client import ClientSession, ClientTimeout, TCPConnector

from .const import LOGGER
from .errors import UnsupportedError
from .models import CombinedModels, Device, Measurement, State, System


class HomeWizardEnergy:
    """Base class for HomeWizard Energy API."""

    _session: ClientSession | None = None
    _close_session: bool = False
    _request_timeout: int = 10
    _host: str

    _device: Device | None = None

    _lock: asyncio.Lock

    def __init__(
        self,
        host: str,
        clientsession: ClientSession = None,
        timeout: int = 10,
    ):
        """Create a HomeWizard Energy object.

        Args:
            host: IP or URL for device.
            clientsession: The clientsession.
            timeout: Request timeout in seconds.
        """
        self._host = host
        self._session = clientsession
        self._close_session = clientsession is None
        self._request_timeout = timeout

        self._lock = asyncio.Lock()

    @property
    def host(self) -> str:
        """Return the hostname of the device.

        Returns:
            host: The used host

        """
        return self._host

    async def combined(self) -> CombinedModels:
        """Get all information."""

        async def fetch_data(coroutine):
            try:
                return await coroutine
            except (UnsupportedError, NotImplementedError):
                return None

        device, measurement, system, state = await asyncio.gather(
            fetch_data(self.device()),
            fetch_data(self.measurement()),
            fetch_data(self.system()),
            fetch_data(self.state()),
        )

        return CombinedModels(
            device=device, measurement=measurement, system=system, state=state
        )

    async def device(self, reset_cache: bool = False) -> Device:
        """Get the device information."""
        raise NotImplementedError

    async def measurement(self) -> Measurement:
        """Get the current measurement."""
        raise NotImplementedError

    async def system(
        self,
        cloud_enabled: bool | None = None,
        status_led_brightness_pct: int | None = None,
        api_v1_enabled: bool | None = None,
    ) -> System:
        """Get/set the system."""
        raise NotImplementedError

    async def state(
        self,
        power_on: bool | None = None,
        switch_lock: bool | None = None,
        brightness: int | None = None,
    ) -> State:
        """Get/set the state."""
        raise UnsupportedError("State is not supported")

    async def identify(
        self,
    ) -> None:
        """Identify the device."""
        raise NotImplementedError

    async def reboot(
        self,
    ) -> None:
        """Reboot the device."""
        raise UnsupportedError("Reboot is not supported")

    async def close(self) -> None:
        """Close client session."""
        LOGGER.debug("Closing clientsession")
        if self._session and self._close_session:
            await self._session.close()

    async def _create_clientsession(self) -> None:
        """Create a client session."""

        LOGGER.debug("Creating clientsession")

        if self._session is not None:
            raise RuntimeError("Session already exists")

        connector = TCPConnector(
            enable_cleanup_closed=True,
            limit_per_host=1,
        )

        self._close_session = True

        self._session = ClientSession(
            connector=connector, timeout=ClientTimeout(total=self._request_timeout)
        )

    async def __aenter__(self) -> HomeWizardEnergy:
        """Async enter.

        Returns:
            The HomeWizardEnergy object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
