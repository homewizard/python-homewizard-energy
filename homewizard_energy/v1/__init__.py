"""Representation of a HomeWizard Energy device."""

from __future__ import annotations

import asyncio
from collections.abc import Callable, Coroutine
from http import HTTPStatus
from typing import Any, TypeVar

import async_timeout
import backoff
from aiohttp.client import ClientError, ClientResponseError
from aiohttp.hdrs import METH_GET, METH_PUT

from ..const import LOGGER
from ..errors import DisabledError, NotFoundError, RequestError, UnsupportedError
from ..homewizard_energy import HomeWizardEnergy
from ..models import Device, Measurement, State, StateUpdate, System, SystemUpdate

T = TypeVar("T")


def optional_method(
    func: Callable[..., Coroutine[Any, Any, T]],
) -> Callable[..., Coroutine[Any, Any, T]]:
    """Check if method is supported."""

    async def wrapper(self, *args, **kwargs) -> T:
        try:
            return await func(self, *args, **kwargs)
        except NotFoundError as ex:
            raise UnsupportedError(f"{func.__name__} is not supported") from ex

    return wrapper


# pylint: disable=abstract-method
class HomeWizardEnergyV1(HomeWizardEnergy):
    """Communicate with a HomeWizard Energy device."""

    async def device(self, reset_cache: bool = False) -> Device:
        """Return the device object."""

        if self._device is not None and not reset_cache:
            return self._device

        _, response = await self._request("api")
        device = Device.from_json(response)

        # Cache device object
        self._device = device
        return device

    async def measurement(self) -> Measurement:
        """Return the data object."""
        _, response = await self._request("api/v1/data")
        return Measurement.from_json(response)

    @optional_method
    async def system(
        self,
        cloud_enabled: bool | None = None,
        status_led_brightness_pct: int | None = None,
        api_v1_enabled: bool | None = None,
    ) -> System:
        """Return the system object."""

        # Legacy: raise on unsupported field
        if api_v1_enabled is not None:
            raise UnsupportedError("Setting api_v1_enabled is not supported in v1")

        # Legacy: route 'status_led_brightness_pct' to state
        if status_led_brightness_pct is not None:
            state = await self.state(brightness=status_led_brightness_pct * 2.55)
            return System(status_led_brightness_pct=state.brightness / 2.55)

        if cloud_enabled is not None:
            # Executing the update
            data = SystemUpdate(cloud_enabled=cloud_enabled).to_dict()
            _, response = await self._request(
                "api/v1/system", method=METH_PUT, data=data
            )

        else:
            _, response = await self._request("api/v1/system")

        system = System.from_json(response)
        return system

    @optional_method
    async def state(
        self,
        power_on: bool | None = None,
        switch_lock: bool | None = None,
        brightness: int | None = None,
    ) -> State:
        """Return or update the state object."""

        if self._device is not None and self._device.supports_state() is False:
            raise UnsupportedError("State is not supported")

        if power_on is not None or switch_lock is not None or brightness is not None:
            data = StateUpdate(
                power_on=power_on, switch_lock=switch_lock, brightness=brightness
            ).to_dict()
            _, response = await self._request(
                "api/v1/state", method=METH_PUT, data=data
            )

        else:
            _, response = await self._request("api/v1/state")

        state = State.from_json(response)
        return state

    @optional_method
    async def identify(
        self,
    ) -> bool:
        """Send identify request."""

        if self._device is not None and self._device.supports_identify() is False:
            raise UnsupportedError("State is not supported")

        await self._request("api/v1/identify", method=METH_PUT)
        return True

    @backoff.on_exception(backoff.expo, RequestError, max_tries=3, logger=None)
    async def _request(
        self, path: str, method: str = METH_GET, data: object = None
    ) -> tuple[HTTPStatus, dict[str, Any] | None]:
        """Make a request to the API."""

        if self._session is None:
            await self._create_clientsession()

        # Construct request
        url = f"http://{self.host}/{path}"
        headers = {"Content-Type": "application/json"}

        LOGGER.debug("%s, %s, %s", method, url, data)

        try:
            async with async_timeout.timeout(self._request_timeout):
                resp = await self._session.request(
                    method,
                    url,
                    json=data,
                    headers=headers,
                )
                LOGGER.debug("%s, %s", resp.status, await resp.text("utf-8"))
        except asyncio.TimeoutError as exception:
            raise RequestError(
                f"Timeout occurred while connecting to the HomeWizard Energy device at {self.host}"
            ) from exception
        except (ClientError, ClientResponseError) as exception:
            raise RequestError(
                f"Error occurred while communicating with the HomeWizard Energy device at {self.host}"
            ) from exception

        match resp.status:
            case HTTPStatus.FORBIDDEN:
                raise DisabledError(
                    "API disabled. API must be enabled in HomeWizard Energy app"
                )
            case HTTPStatus.NOT_FOUND:
                raise NotFoundError("Resource not found")
            case HTTPStatus.OK:
                pass

        if resp.status != HTTPStatus.OK:
            # Something else went wrong
            raise RequestError(f"API request error ({resp.status})")

        return (resp.status, await resp.text())

    async def __aenter__(self) -> HomeWizardEnergyV1:
        """Async enter.

        Returns:
            The HomeWizardEnergyV1 object.
        """
        return self
