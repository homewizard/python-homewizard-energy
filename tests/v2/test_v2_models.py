"""Test for HomeWizard Energy Models."""

import json

import pytest
from syrupy.assertion import SnapshotAssertion

from homewizard_energy.models import Device, Measurement, System, SystemUpdate

from . import load_fixtures

pytestmark = [pytest.mark.asyncio]


@pytest.mark.parametrize(
    ("model", "fixtures"),
    [
        ("HWE-P1", ["device"]),
        ("HWE-BAT", ["device"]),
    ],
)
async def test_device(model: str, fixtures: str, snapshot: SnapshotAssertion):
    """Test Device model."""
    for fixture in fixtures:
        data = Device.from_dict(json.loads(load_fixtures(f"{model}/{fixture}.json")))
        assert data

        assert snapshot == data


@pytest.mark.parametrize(
    ("model", "fixtures"),
    [
        (
            "HWE-P1",
            [
                "measurement_1_phase_no_gas",
                "measurement_3_phase_with_gas_with_watermeter",
                "measurement_invalid_ean",
                "measurement_invalid_external",
            ],
        ),
        (
            "HWE-BAT",
            [
                "measurement",
            ],
        ),
    ],
)
async def test_measurement(model: str, fixtures: str, snapshot: SnapshotAssertion):
    """Test Measurement model."""
    for fixture in fixtures:
        data = Measurement.from_dict(
            json.loads(load_fixtures(f"{model}/{fixture}.json"))
        )
        assert data

        assert snapshot == data


async def test_measurement_ignores_invalid_tariff():
    """Test Measurement model ignores invalid tariff."""

    measurement = Measurement.from_dict({"tariff": 5432})
    assert measurement
    assert measurement.tariff is None


@pytest.mark.parametrize(
    ("model", "fixtures"),
    [
        ("HWE-P1", ["system"]),
        ("HWE-BAT", ["system"]),
    ],
)
async def test_system(model: str, fixtures: str, snapshot: SnapshotAssertion):
    """Test System model."""
    for fixture in fixtures:
        data = System.from_dict(json.loads(load_fixtures(f"{model}/{fixture}.json")))
        assert data

        assert snapshot == data


@pytest.mark.parametrize(
    ("cloud_enabled", "status_led_brightness_pct", "api_v1_enabled"),
    [
        (True, None, None),
        (None, 100, None),
        (None, None, True),
        (True, 100, None),
        (True, None, True),
        (None, 100, True),
        (True, 100, True),
    ],
)
async def test_system_update(
    cloud_enabled: bool,
    status_led_brightness_pct: int,
    api_v1_enabled: bool,
    snapshot: SnapshotAssertion,
):
    """Test System update."""
    data = SystemUpdate(
        cloud_enabled=cloud_enabled,
        status_led_brightness_pct=status_led_brightness_pct,
        api_v1_enabled=api_v1_enabled,
    )
    assert snapshot == data.to_dict()


async def test_system_update_raises_when_none_set():
    """Test systemUpdate raises when all values are None."""
    update = SystemUpdate(
        cloud_enabled=None, status_led_brightness_pct=None, api_v1_enabled=None
    )

    with pytest.raises(ValueError):
        update.to_dict()


@pytest.mark.parametrize(
    ("model", "supports_state", "supports_identify", "supports_cloud_enable"),
    [
        ("HWE-P1", False, True, True),
        ("HWE-BAT", False, True, False),
    ],
)
async def test_device_support_functions(
    model: str,
    supports_state: bool,
    supports_identify: bool,
    supports_cloud_enable: bool,
):
    """Test Device model support functions."""
    device = Device.from_dict(json.loads(load_fixtures(f"{model}/device.json")))
    assert device

    assert device.supports_state() == supports_state
    assert device.supports_identify() == supports_identify
    assert device.supports_cloud_enable() == supports_cloud_enable
    assert device.supports_reboot() is True  # Always True for v2
