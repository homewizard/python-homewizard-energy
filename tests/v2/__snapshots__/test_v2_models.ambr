# serializer version: 1
# name: test_batteries[HWE-P1-fixtures0]
  Batteries(mode=<Mode.ZERO: 'zero'>, power_w=-404.0, target_power_w=-400.0, max_consumption_w=1600.0, max_production_w=800.0)
# ---
# name: test_batteries_update[standby]
  dict({
    'mode': 'standby',
  })
# ---
# name: test_batteries_update[to_full]
  dict({
    'mode': 'to_full',
  })
# ---
# name: test_batteries_update[zero]
  dict({
    'mode': 'zero',
  })
# ---
# name: test_device[HWE-BAT-fixtures1]
  Device(model_name='Plug-In Battery', id='appliance/battery/5c2fafaabbcc', product_name='Plug-in Battery', product_type='HWE-BAT', serial='5c2fafaabbcc', api_version=<AwesomeVersion SemVer '2.0.0'>, firmware_version='6.00')
# ---
# name: test_device[HWE-P1-fixtures0]
  Device(model_name='Wi-Fi P1 Meter', id='appliance/p1dongle/5c2fafaabbcc', product_name='P1 Meter', product_type='HWE-P1', serial='5c2fafaabbcc', api_version=<AwesomeVersion SemVer '2.0.0'>, firmware_version='6.00')
# ---
# name: test_measurement[HWE-BAT-fixtures1]
  Measurement(wifi_ssid=None, wifi_strength=None, energy_import_kwh=123.456, energy_import_t1_kwh=None, energy_import_t2_kwh=None, energy_import_t3_kwh=None, energy_import_t4_kwh=None, energy_export_kwh=321.654, energy_export_t1_kwh=None, energy_export_t2_kwh=None, energy_export_t3_kwh=None, energy_export_t4_kwh=None, power_w=1234.0, power_l1_w=None, power_l2_w=None, power_l3_w=None, voltage_v=230.123, voltage_l1_v=None, voltage_l2_v=None, voltage_l3_v=None, current_a=5.678, current_l1_a=None, current_l2_a=None, current_l3_a=None, apparent_power_va=None, apparent_power_l1_va=None, apparent_power_l2_va=None, apparent_power_l3_va=None, reactive_power_var=None, reactive_power_l1_var=None, reactive_power_l2_var=None, reactive_power_l3_var=None, power_factor=None, power_factor_l1=None, power_factor_l2=None, power_factor_l3=None, frequency_hz=50.123, timestamp=None, protocol_version=None, meter_model=None, unique_id=None, tariff=None, voltage_sag_l1_count=None, voltage_sag_l2_count=None, voltage_sag_l3_count=None, voltage_swell_l1_count=None, voltage_swell_l2_count=None, voltage_swell_l3_count=None, any_power_fail_count=None, long_power_fail_count=None, average_power_15m_w=None, monthly_power_peak_w=None, monthly_power_peak_timestamp=None, external_devices=None, active_liter_lpm=None, total_liter_m3=None, cycles=123, state_of_charge_pct=50.123)
# ---
# name: test_measurement[HWE-P1-fixtures0]
  Measurement(wifi_ssid=None, wifi_strength=None, energy_import_kwh=None, energy_import_t1_kwh=10830.511, energy_import_t2_kwh=2948.827, energy_import_t3_kwh=None, energy_import_t4_kwh=None, energy_export_kwh=None, energy_export_t1_kwh=1285.951, energy_export_t2_kwh=2876.51, energy_export_t3_kwh=None, energy_export_t4_kwh=None, power_w=-678.0, power_l1_w=-676.0, power_l2_w=None, power_l3_w=None, voltage_v=None, voltage_l1_v=None, voltage_l2_v=None, voltage_l3_v=None, current_a=None, current_l1_a=None, current_l2_a=None, current_l3_a=None, apparent_power_va=None, apparent_power_l1_va=None, apparent_power_l2_va=None, apparent_power_l3_va=None, reactive_power_var=None, reactive_power_l1_var=None, reactive_power_l2_var=None, reactive_power_l3_var=None, power_factor=None, power_factor_l1=None, power_factor_l2=None, power_factor_l3=None, frequency_hz=None, timestamp=datetime.datetime(2024, 6, 28, 14, 12, 34), protocol_version=50, meter_model='ISKRA  2M550T-101', unique_id='Never Gonna Give You Up', tariff=None, voltage_sag_l1_count=None, voltage_sag_l2_count=None, voltage_sag_l3_count=None, voltage_swell_l1_count=None, voltage_swell_l2_count=None, voltage_swell_l3_count=None, any_power_fail_count=None, long_power_fail_count=None, average_power_15m_w=None, monthly_power_peak_w=None, monthly_power_peak_timestamp=None, external_devices=None, active_liter_lpm=None, total_liter_m3=None, cycles=None, state_of_charge_pct=None)
# ---
# name: test_measurement[HWE-P1-fixtures0].1
  Measurement(wifi_ssid=None, wifi_strength=None, energy_import_kwh=13779.338, energy_import_t1_kwh=10830.511, energy_import_t2_kwh=2948.827, energy_import_t3_kwh=None, energy_import_t4_kwh=None, energy_export_kwh=0.0, energy_export_t1_kwh=0.0, energy_export_t2_kwh=0.0, energy_export_t3_kwh=None, energy_export_t4_kwh=None, power_w=-543.0, power_l1_w=-676.0, power_l2_w=133.0, power_l3_w=0.0, voltage_v=None, voltage_l1_v=None, voltage_l2_v=None, voltage_l3_v=None, current_a=6.0, current_l1_a=-4.0, current_l2_a=2.0, current_l3_a=0.0, apparent_power_va=None, apparent_power_l1_va=None, apparent_power_l2_va=None, apparent_power_l3_va=None, reactive_power_var=None, reactive_power_l1_var=None, reactive_power_l2_var=None, reactive_power_l3_var=None, power_factor=None, power_factor_l1=None, power_factor_l2=None, power_factor_l3=None, frequency_hz=None, timestamp=datetime.datetime(2024, 6, 28, 14, 12, 34), protocol_version=50, meter_model='ISKRA  2M550T-101', unique_id='Never Gonna Let You Down', tariff=2, voltage_sag_l1_count=1, voltage_sag_l2_count=1, voltage_sag_l3_count=0, voltage_swell_l1_count=0, voltage_swell_l2_count=0, voltage_swell_l3_count=0, any_power_fail_count=4, long_power_fail_count=5, average_power_15m_w=123.0, monthly_power_peak_w=1111.0, monthly_power_peak_timestamp=datetime.datetime(2024, 6, 4, 10, 11, 22), external_devices={'gas_meter_Never gonna run around': ExternalDevice(unique_id='Never gonna run around', type=<DeviceType.GAS_METER: 'gas_meter'>, value=2569.646, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 0)), 'water_meter_and desert you': ExternalDevice(unique_id='and desert you', type=<DeviceType.WATER_METER: 'water_meter'>, value=123.456, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 5))}, active_liter_lpm=None, total_liter_m3=None, cycles=None, state_of_charge_pct=None)
# ---
# name: test_measurement[HWE-P1-fixtures0].2
  Measurement(wifi_ssid=None, wifi_strength=None, energy_import_kwh=None, energy_import_t1_kwh=None, energy_import_t2_kwh=None, energy_import_t3_kwh=None, energy_import_t4_kwh=None, energy_export_kwh=None, energy_export_t1_kwh=None, energy_export_t2_kwh=None, energy_export_t3_kwh=None, energy_export_t4_kwh=None, power_w=None, power_l1_w=None, power_l2_w=None, power_l3_w=None, voltage_v=None, voltage_l1_v=None, voltage_l2_v=None, voltage_l3_v=None, current_a=None, current_l1_a=None, current_l2_a=None, current_l3_a=None, apparent_power_va=None, apparent_power_l1_va=None, apparent_power_l2_va=None, apparent_power_l3_va=None, reactive_power_var=None, reactive_power_l1_var=None, reactive_power_l2_var=None, reactive_power_l3_var=None, power_factor=None, power_factor_l1=None, power_factor_l2=None, power_factor_l3=None, frequency_hz=None, timestamp=None, protocol_version=None, meter_model=None, unique_id='\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', tariff=None, voltage_sag_l1_count=None, voltage_sag_l2_count=None, voltage_sag_l3_count=None, voltage_swell_l1_count=None, voltage_swell_l2_count=None, voltage_swell_l3_count=None, any_power_fail_count=None, long_power_fail_count=None, average_power_15m_w=None, monthly_power_peak_w=None, monthly_power_peak_timestamp=None, external_devices={'gas_meter_\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': ExternalDevice(unique_id='\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', type=<DeviceType.GAS_METER: 'gas_meter'>, value=111.111, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 0)), 'heat_meter_\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': ExternalDevice(unique_id='\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', type=<DeviceType.HEAT_METER: 'heat_meter'>, value=222.222, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 0)), 'warm_water_meter_\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': ExternalDevice(unique_id='\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', type=<DeviceType.WARM_WATER_METER: 'warm_water_meter'>, value=333.333, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 0)), 'water_meter_\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': ExternalDevice(unique_id='\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', type=<DeviceType.WATER_METER: 'water_meter'>, value=444.444, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 0)), 'inlet_heat_meter_\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': ExternalDevice(unique_id='\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', type=<DeviceType.INLET_HEAT_METER: 'inlet_heat_meter'>, value=555.555, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 0)), 'None_\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': ExternalDevice(unique_id='\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', type=None, value=555.555, unit=',3', timestamp=datetime.datetime(2024, 6, 28, 14, 0))}, active_liter_lpm=None, total_liter_m3=None, cycles=None, state_of_charge_pct=None)
# ---
# name: test_measurement[HWE-P1-fixtures0].3
  Measurement(wifi_ssid=None, wifi_strength=None, energy_import_kwh=None, energy_import_t1_kwh=None, energy_import_t2_kwh=None, energy_import_t3_kwh=None, energy_import_t4_kwh=None, energy_export_kwh=None, energy_export_t1_kwh=None, energy_export_t2_kwh=None, energy_export_t3_kwh=None, energy_export_t4_kwh=None, power_w=None, power_l1_w=None, power_l2_w=None, power_l3_w=None, voltage_v=None, voltage_l1_v=None, voltage_l2_v=None, voltage_l3_v=None, current_a=None, current_l1_a=None, current_l2_a=None, current_l3_a=None, apparent_power_va=None, apparent_power_l1_va=None, apparent_power_l2_va=None, apparent_power_l3_va=None, reactive_power_var=None, reactive_power_l1_var=None, reactive_power_l2_var=None, reactive_power_l3_var=None, power_factor=None, power_factor_l1=None, power_factor_l2=None, power_factor_l3=None, frequency_hz=None, timestamp=None, protocol_version=None, meter_model=None, unique_id=None, tariff=None, voltage_sag_l1_count=None, voltage_sag_l2_count=None, voltage_sag_l3_count=None, voltage_swell_l1_count=None, voltage_swell_l2_count=None, voltage_swell_l3_count=None, any_power_fail_count=None, long_power_fail_count=None, average_power_15m_w=None, monthly_power_peak_w=None, monthly_power_peak_timestamp=None, external_devices={'None_Never gonna run around': ExternalDevice(unique_id='Never gonna run around', type=None, value=555.555, unit=',3', timestamp=datetime.datetime(2024, 6, 28, 14, 0))}, active_liter_lpm=None, total_liter_m3=None, cycles=None, state_of_charge_pct=None)
# ---
# name: test_system[HWE-BAT-fixtures1]
  System(wifi_strength_pct=46, wifi_ssid='My Wi-Fi', wifi_rssi_db=-77, cloud_enabled=False, uptime_s=356, status_led_brightness_pct=100, api_v1_enabled=None)
# ---
# name: test_system[HWE-P1-fixtures0]
  System(wifi_strength_pct=46, wifi_ssid='My Wi-Fi', wifi_rssi_db=-77, cloud_enabled=False, uptime_s=356, status_led_brightness_pct=100, api_v1_enabled=True)
# ---
# name: test_system_update[None-100-None]
  dict({
    'status_led_brightness_pct': 100,
  })
# ---
# name: test_system_update[None-100-True]
  dict({
    'api_v1_enabled': True,
    'status_led_brightness_pct': 100,
  })
# ---
# name: test_system_update[None-None-True]
  dict({
    'api_v1_enabled': True,
  })
# ---
# name: test_system_update[True-100-None]
  dict({
    'cloud_enabled': True,
    'status_led_brightness_pct': 100,
  })
# ---
# name: test_system_update[True-100-True]
  dict({
    'api_v1_enabled': True,
    'cloud_enabled': True,
    'status_led_brightness_pct': 100,
  })
# ---
# name: test_system_update[True-None-None]
  dict({
    'cloud_enabled': True,
  })
# ---
# name: test_system_update[True-None-True]
  dict({
    'api_v1_enabled': True,
    'cloud_enabled': True,
  })
# ---
